### ISSUE TITLE
Fails after waking from sleep

### ISSUE THREAD

**Issue by sachamasry** [rvaiya/warpd#185]
Running warpd as per the instructions, and build from source, on macOS, has one problem: when the os wakes from sleep, the activation key combination no longer activates warpd, instead resulting in a default os chime only.

I've tried this in foreground and daemon modes with the same result, warpd has to be quit/killed, and restarted.

I'm on macOS Monterrey, 12.6.

Have you come across this problem on macOS or elsewhere?

**Comment by rvaiya** [comment]
Are you suggesting that the activation key ceases to work after closing your laptop lid? I can't seem to reproduce this. You should only need to start it on boot. What version are you running? (`warpd -v`)

**Comment by sachamasry** [comment]
Hi, I'm running the latest, 1.3.4, built from source. Having used it for a few days more, I can see that sometimes the activation key works fine after waking, but other times it simply does nothing. And, I can confirm that warpd is still running.

I'll try to narrow this down to a definitely reproducible issue, once I have more time with it.

**Comment by sachamasry** [comment]
I've run the latest commit and have had three sets of results coming out of sleep: (a) it works fine; (b) it doesn't respond to activation keys at all; and once (c) warpd is activated, the pointer is hidden, and the keyboard is hijacked, but the red dot is not visible, and pressing Esc doesn't bring me out of warpd's control until some keypress resulted in its dismissal.

So, it's not working reliably yet, though it may easily be due to any of the other services I have running on the system which are conflicting with it.

**Comment by rvaiya** [comment]
Thanks for the feedback. I will investigate. I don't daily drive macos, so there are liable to be some bugs in the platform logic.

**Comment by sachamasry** [comment]
Hi, it's happened again that I wake from sleep and warpd is activated, with the mouse invisible and the keyboard fully captured. I was completely unable to do anything to 'get out' of warpd to regain control and kill it from the terminal, and had to forcefully shut down the machine.
Is there a single key or combination to get out of warpd's active mode?

**Comment by rvaiya** [comment]
> Is there a single key or combination to get out of warpd's active mode?

Aside from the key you mapped? No. It probably wouldn't be very helpful since this is likely a bug in the input logic.

I've just made the macos input handling a bit more robust. I believe the root cause of this should be addressed in the latest commit. Can you give it a go?

**Comment by sachamasry** [comment]
I have just compiled the latest commit and a slew of problems have just started.

First, the 'Accessibility Access' popup now incessantly keeps popping up, regardless of System Preferences settings, or klicking the 'Deny' button (see below). There is absolutely no way to dismiss this popup, as it respawns every few seconds!

<img width="464" alt="Screen Shot 2022-10-23 at 12 20 31" src="https://user-images.githubusercontent.com/28275251/197406333-b6d8a766-1615-4df6-90ca-4032ebaf5ab8.png">
<img width="464" alt="Screen Shot 2022-10-23 at 12 23 53" src="https://user-images.githubusercontent.com/28275251/197406337-690cd568-d85a-46c4-9ada-5146ba94dd0b.png">

Secondly, I am not running warpd in a shell, nor as a 'Login Item' in my user profile, so I can't see why this modal is constantly being triggered. Running `pgrep -l warp` shows no processes matching, so it should not be running even!

Thirdly, following the instructions that "The service can be disabled with `launchctl unload /Library/LaunchAgents/com.warpd.warpd`" doesn't work, resulting in the following error:

`❯ launchctl unload /Library/LaunchAgents/com.warpd.warpd
Unload failed: 5: Input/output error
Try running `launchctl bootout` as root for richer errors.`

Similarly, `launchctl load /Library/LaunchAgents/com.warpd.warpd.plist` results in something similar:

`❯ launchctl load /Library/LaunchAgents/com.warpd.warpd.plist
Load failed: 5: Input/output error
Try running `launchctl bootstrap` as root for richer errors.`

All in all, it's now become a bit of a systemic problem.

**Comment by sachamasry** [comment]
Ok, I found out how to resolve (3):

`launchctl disable gui/502/com.warpd.warpd
launchctl bootout gui/502/com.warpd.warpd`

where `502` is my $UID.

That's taken me out of System Preferences neverending popup hell, and I'm back to running warpd, in the terminal, in foreground mode, until further update:

`wardp -f`

**Comment by rvaiya** [comment]
You probably need to run `tccutil reset Accessibility` (mentioned in the readme). macos doesn't seem to properly associate the checksum rather than the binary with the requisite permissions, which it seems to cache.

**Comment by rvaiya** [comment]
> Thirdly, following the instructions that "The service can be disabled with launchctl unload /Library/LaunchAgents/com.warpd.warpd" doesn't work, resulting in the following error:

Sorry, that should have been ` launchctl unload /Library/LaunchAgents/com.warpd.warpd.plist`. I have updated the readme accordingly.

**Comment by sachamasry** [comment]
As per the `launchctl` manual, the targets `load` and `unload` have been deprecated. The advice is to use `bootstrap`/`bootout` and `enable`/`disable`, so the README instructions don't work for me.

I have tried to use these alternate targets, but due to my lack of launchctl knowledge, have been unable to start it using LaunchAgent.

However, I have run warpd in the terminal (`warpd -f`) every day since your last message, and it appears to work properly with none of the issues I described before. In other words, warpd seems to work perfectly every time the computer now comes out of sleep. 

Thank you very much!

**Comment by rvaiya** [comment]
Thanks for the confirmation, however I would like to work out any bugs in the shipped launchctl service. Can you clarify what you mean by 'unable to start it'? `launchctl load /Library/LaunchAgents/com.warpd.warpd.plist` should suffice for installing the service.

**Comment by sachamasry** [comment]
Ok, I don't know what did it, but your instruction above (like in the README) now worked, without error. Previously, I was seeing this for `unload` (and something similar for `load`):

`❯ launchctl unload /Library/LaunchAgents/com.warpd.warpd.plist`
`Unload failed: 5: Input/output error`
`Try running `launchctl bootout` as root for richer errors.`

The man page says they are legacy subcommands. But I think this is all irrelevant, as I now believe multiple instances were running through my bad issuing of launchctl commands, so the above error messages are irrelevant and your documentation is correct and simply works.

However, I have had two instances today of (presumably) warpd completely locking out the keyboard, to the point where I couldn't use the computer at all, and had to log in remotely through `ssh` to `kill` or `unload` warpd to get back in. As I had not experienced this before, I will assume that it is down to the above-described conflict (multiple instances running).

Why did I end up with multiple instances? It seems I created the problem by issuing the `launchctl` commands in `tmux`. A 'superuser.com' answer (https://superuser.com/questions/860735/killing-a-launchctl-process-cassandra) suggests that running launchctl in tmux is problematic. As soon as I ran the commands in a plain terminal, everything works!
